### ISSUE TITLE
Exception thrown by a callback function: Trying to cast a lua variable from "string" to "NSt15__exception_ptr13exception_ptrE" (pdns.DROP causing havoc)

### ISSUE THREAD

**Issue by grinapo** [PowerDNS/pdns#11287]
> This bugreport is created to make this problem searchable. 

 - Program: Recursor
 - Issue type: bugreport

Upgrading recursor from 4.3.xx to 4.6.xx breaks lua script in a most unintuitive way. It is really hard to debug, really hard to guess at the problem and the documentation is lying. 

### What happens after upgrade
If you have a lua script you may get infinite amounts of:

```
STL Error( ... Exception thrown by a callback function: Trying to cast a lua variable from "string" to "NSt15__exception_ptr13exception_ptrE" ...)
```
With no visible reason or explanation, nor any line numbers or function indicators. "It is in the lua, any of them, somewhere."

### Cause
`dq.rcode = pdns.DROP` _seems to be_ depreciated, then transformed into a bug generator. This assignment actually generated the error message, and may or may not cause other problems.

The documentation suggests that _this is a proper solution_. 

By browsing issues and searching for wildly random terms it turns out that issue #10080 and pull #10111 actually took care of something, and suggested a documentation change which may or may not actually happen (but definitely have not changed the examples section in the lua titles of the recursor documentation), and in due time intentionally or not broke `pdns.DROP` with little visibility.

### Solution
Code like
```lua
function preresolve(dq)
  if blocklist:check(dq.qname) then
    dq.rcode=pdns.DROP
    return true
  end
end
```
shall be replaced by
```lua
function preresolve(dq)
  if blocklist:check(dq.qname) then
    dq.appliedPolicy.policyKind = pdns.policykinds.Drop
    return false
  end
end
```
and the error will go away.

### Suggestions
- fix the doc
- fix the error message (I guess `pdns.DROP` is `nil`now, or it's maybe a non-existing element in some collection?)
- warn the people, it hurts.
- feel free to close this issue, if you believe there's nothing more left to do.

**Comment by omoerbeek** [comment]
- If you see sport where this is documented in a wrong way please specify where
- There's documentation of the new behaviour here:https://docs.powerdns.com/recursor/lua-scripting/hooks.html#semantics
- The error message must be made more clear.
- An extra warning in the `preresolve` docs is certainly needed
- A note in the upgrade guide should also be added

**Comment by omoerbeek** [comment]
I believe all issues mentioned are addressed in #11288

**Comment by grinapo** [comment]
Thank you, as far as I see all problems have been addressed.
